[USING] WalletGeneration.md : Correct factual errors, and move some info to new/advanced password chapter (#593)
zkSNACKs/WasabiDoc: docs/using-wasabi/PasswordBestPractices.md

---
{
  "title": "Password Best Practices",
  "description": "A detailed guide about how to create and evaluate a strong password. This is the Wasabi documentation, an archive of knowledge about the open-source, non-custodial and privacy-focused Bitcoin wallet for desktop."
}
---

# Password best practices

:::tip
Don't "roll your own" crypto.
:::

You shouldn't try to re-invent the wheel for something as complicated and nuanced as cryptography, and especially in regards to entropy.

These are some of the industry best practices:

[[toc]]

## Use Diceware

Rolling dice is an easy and effective way to get high entropy and randomness in a password that can even be memorized.

The Diceware method is a great strategy for your most precious passwords (e.g. the password to your computer, your backups, or your encryption key).
You can even use Diceware to create secure wallets.

> To generate a password using Diceware, you just need a good die, a pen, and some paper.

Author: ```suggestion
To generate a password using Diceware, you just need a high-quality die, a pen, and some paper.
```

After six sets of five rolls, we ended up with `52611` `51631` `63432` `43123` `21641` and `14146`.
This corresponds to the password `salvo rhoda walton mudd croft bride`.

This six-word passphrase, even though easy to remember, is just one of 2.21 x 10^23, or 221 sextillion possible six-word combinations taken from the list of 7776 words.
A computer capable of 1 trillion guesses/second would take, on average, over 3500 years to correctly guess this or any six-word combination of words from the list, if they are chosen randomly.

A 6-word passphrase derived as just discussed has 77 [bits of entropy](/using-wasabi/PasswordBestPractices.md#how-to-calculate-entropy).
Best practices for being secure against any adversary [currently recommend](https://blog.securityevaluators.com/understanding-password-complexity-5e0d23643a2a) a minimum of near 80 bits of entropy, so this example phrase barely qualifies.

## Use a dictionary

All of the major languages have at least 100,000 words in their dictionaries.
Most have over 300,000 words.
You can take a dictionary in any language that you are comfortable with and create a very strong password by randomly selecting as few as 5 words from it.

> The quick, but not technically random way would be to flip open your chosen dictionary to any seemingly random page, close your eyes, and put your fingertip down on the page. Choose the word nearest your fingertip that has at least 4 letters.

Reviewer: ```suggestion
The quick, but not technically random way would be to flip open your chosen dictionary to any seemingly random page, close your eyes, and put your fingertip down on the page.
Choose the word nearest your fingertip that has at least 4 letters.
```

It is up to the user to choose between a password manager that stores its data in the cloud, making it easy to stay in sync across various devices, or one that only stores its data locally.

As it so often happens, more convenience means less security in this decision.

## Password entropy

"Entropy" is a term commonly used to define password strength.
It is typically used to describe the combined effect of randomness, length, and number of characters (lower-case/upper-case letters, numbers, and special characters) or words in a password or passphrase.

Calculating entropy allows the strength of different randomly constructed passwords or phrases to be compared.
It also defines the size of the search-space an attacker would need to cover to find the password by trying all possibilities.

### How to calculate entropy

Assuming a password or passphrase is constructed randomly, its level of entropy can be calculated in this way:

- For a password, the total number of possible outcomes given a specific length (L), and number of characters used (C), is:
C raised to the power of L, or C^L.<br>

> Example: `jwodnrosqf` is lower-case only (C = 26) and 10 characters long (L = 10). So, 26^10 = 141,000,000,000,000 (141 trillion).<br><br>

Reviewer: ```suggestion
Example: `jwodnrosqf` is lower-case only (C = 26) and 10 characters long (L = 10).
So, 26^10 = 141,000,000,000,000 (141 trillion).<br><br>
```

Including upper-case letters in the same password: `jWoDNrosqF` to double (C), results in 52^10 possible passwords = 145,000,000,000,000,000 (145 quadrillion).<br><br>
Including special characters in the same password: `jW0DNro$qF` raises (C) to 95, resulting in 95^10 possibilities = 59,900,000,000,000,000,000 (59.9 quintillion).

- For a passphrase chosen randomly from a list of words, the total number of possible outcomes given a specific number of words (N) and wordlist size (S) is: S raised to the power of N, or S^N.<br>
Example: The EFF Diceware list contains 7776 words.
Randomly selecting 5 words from the list results in 7776^5 possible passphrases = 28,400,000,000,000,000,000 (28.4 quintillion).

Just knowing how to do these basic calculations allows you to compare the strength of different randomly-generated passwords or phrases.

### "Bits" of entropy

Password entropy is often discussed and compared in a context of `bits of entropy`.
This is, of course, because computers can only process 0's and 1's, or bits, so everything related is expressed that way.

> To calculate bits of entropy, use a [calculator that does logarithms](https://miniwebtool.com/log-base-2-calculator/?num1=7776) and compute log2(x), where x = the total number of possible outcomes for an event.

Reviewer: ```suggestion
To calculate bits of entropy, use a [calculator that does logarithms](https://miniwebtool.com/log-base-2-calculator/) and compute log2(x), where x = the total number of possible outcomes for an event.
```

From that, we can now calculate that, in order to get the same level of entropy with a password instead of a 5-word passphrase, the password must be at least 10 characters in length.

### How much entropy is needed

Entropy matters because it allows one to compare the required `search-space` of a potential password with the `search-speed` known to be availble to various potential adversaries.
In doing so, it is possible to then know how long it would take an adversary with known resources to attempt every mathematically possible password or passphrase for a given level of entropy.

Edward Snowden revealed in 2013 that the NSA had the ability to guess passwords at a rate of 1 trillion guesses/second.
At that speed of guessing, the password in the above example (28.4 quintillion possibilities) could be guessed in an average of 5 months.

Since there is an equal chance of finding the password in the first half of the total number of guesses as there is in finding it in the second half of the total number, it is accurate to assume that on average, a password will be found in 1/2 of the time required to do an exhaustive search.

> That means just over 11 months for the example just mentioned.

Reviewer: ```suggestion
That means 2 months and a half for the example just mentioned.
```

---
{
  "title": "Password Best Practices",
  "description": "A detailed guide about how to create and evaluate a strong password. This is the Wasabi documentation, an archive of knowledge about the open-source, non-custodial and privacy-focused Bitcoin wallet for desktop."
}
---

# Password best practices

:::tip
Don't "roll your own" crypto.
:::

You shouldn't try to re-invent the wheel for something as complicated and nuanced as cryptography, and especially in regards to entropy.

These are some of the industry best practices:

[[toc]]

## Use Diceware

> Rolling dice is an easy and effective way to get high entropy and randomness in a password that can even be memorized.

Reviewer: IMO we shouldn't mention memorizing random passwords with high entropy.
```suggestion
Rolling dice is an easy and effective way to get high entropy and randomness in a password.
```

Author: Memorizing a diceware password is a feature, not a bug. This is a good article:
https://theintercept.com/2015/03/26/passphrases-can-memorize-attackers-cant-guess/

The Diceware method is a great strategy for your most precious passwords (e.g. the password to your computer, your backups, or your encryption key).
You can even use Diceware to create secure wallets.

> To generate a password using Diceware, you just need a good die, a pen, and some paper.

Reviewer: we should mention that you should have a very good die.

Author: Done!

> You can find the diceware list at the [Electronic Frontier Foundation website](https://www.eff.org/deeplinks/2016/07/new-wordlists-random-passphrases).

Reviewer: ```suggestion
You can find EFF's Diceware list at the [Electronic Frontier Foundation website](https://www.eff.org/deeplinks/2016/07/new-wordlists-random-passphrases).
```

Make sure you are alone and there are no cameras nearby.
For maximum protection, disconnect your computer from the internet (after you save the Diceware list) and cover your webcam.

To start, roll the die five times.
Record the number from each roll using the pen and paper.
You will end up with a five-digit number.
We got `52611`.
Now search on the Diceware list for the five digit number you just created.
Write down the word the number corresponds to, in our case it is `salvo`.
This word by itself is not a good password, it would only take about a thousandth of a second to crack.
Repeat the dice rolling process at least five more times.

After six sets of five rolls, we ended up with `52611` `51631` `63432` `43123` `21641` and `14146`.
This corresponds to the password `salvo rhoda walton mudd croft bride`.

> This six-word passphrase, even though easy to remember, is just one of 2.21 x 10^23, or 221 sextillion possible six-word combinations taken from the list of 7776 words.

Reviewer: IMO we shouldn't mention memorizing random passwords.
```suggestion
This six-word passphrase is just one of 2.21 x 10^23, or 221 sextillion possible six-word combinations taken from the list of 7776 words.
```

Author: Memorizing a diceware password is a feature, not a bug. This is a good article:
https://theintercept.com/2015/03/26/passphrases-can-memorize-attackers-cant-guess/

A computer capable of 1 trillion guesses/second would take, on average, over 3500 years to correctly guess this or any six-word combination of words from the list, if they are chosen randomly.

A 6-word passphrase derived as just discussed has 77 [bits of entropy](/using-wasabi/PasswordBestPractices.md#how-to-calculate-entropy).
Best practices for being secure against any adversary [currently recommend](https://blog.securityevaluators.com/understanding-password-complexity-5e0d23643a2a) a minimum of near 80 bits of entropy, so this example phrase barely qualifies.

## Use a dictionary

> All of the major languages have at least 100,000 words in their dictionaries. Most have over 300,000 words.

Reviewer: ```suggestion
All of the major languages have at least 100,000 words in their dictionaries.
Most have over 300,000 words.
```

> You can take a dictionary in any language that you are comfortable with and create a very strong password by randomly selecting as few as 5 words from it.

Reviewer: I think this is a wrong advice.
I am not an expert but I think the words used to generate passwords are carefully selected and the number of words and their order is known.
While using any dictionary is bad because the list of words may change.

Author: The words used to create the seed are specifically chosen and are standardized across wallets, and will not work in all combinations.
Here, we are talking about a wallet password to secure the seed words. You can use anything or any words you want for a password

It also defines the size of the search-space an attacker would need to cover to find the password by trying all possibilities.

### How to calculate entropy

Assuming a password or passphrase is constructed randomly, its level of entropy can be calculated in this way:

- For a password, the total number of possible outcomes given a specific length (L), and number of characters used (C), is:
C raised to the power of L, or C^L.<br>
Example: `jwodnrosqf` is lower-case only (C = 26) and 10 characters long (L = 10). So, 26^10 = 141,000,000,000,000 (141 trillion).<br><br>
Including upper-case letters in the same password: `jWoDNrosqF` to double (C), results in 52^10 possible passwords = 145,000,000,000,000,000 (145 quadrillion).<br><br>
Including special characters in the same password: `jW0DNro$qF` raises (C) to 95, resulting in 95^10 possibilities = 59,900,000,000,000,000,000 (59.9 quintillion).

- For a passphrase chosen randomly from a list of words, the total number of possible outcomes given a specific number of words (N) and wordlist size (S) is: S raised to the power of N, or S^N.<br>

> Example: The EFF diceware wordlist contains 7776 words.

Reviewer: ```suggestion
Example: The EFF Diceware list contains 7776 words.
```

So, a 5-word passphrase selected randomly from the EFF list will have 5 x 12.92 = 64 bits of entropy.

Since the combination of: lower + uppercase letters, numbers, and special characters = 95, a password constructed using all of these will have log2(95), or 6.57 bits of entropy for each character that comprises the password.

From that, we can now calculate that, in order to get the same level of entropy with a password instead of a 5-word passphrase, the password must be at least 10 characters in length.

### How much entropy is needed

Entropy matters because it allows one to compare the required `search-space` of a potential password with the `search-speed` known to be availble to various potential adversaries.
In doing so, it is possible to then know how long it would take an adversary with known resources to attempt every mathematically possible password or passphrase for a given level of entropy.

Edward Snowden revealed in 2013 that the NSA had the ability to guess passwords at a rate of 1 trillion guesses/second.

> At that speed of guessing, the password in the above example (28.4 quintillion possibilities) could be guessed in at most, 11 months..

Author: ```suggestion
At that speed of guessing, the password in the above example (28.4 quintillion possibilities) could be guessed in at most, 11 months.
```

> At that speed of guessing, the password in the above example (28.4 quintillion possibilities) could be guessed in an average of 5 months.

Author: ```suggestion
At that speed of guessing, the password in the above example (28.4 quintillion possibilities) could be guessed in at most, 11 months..
```
This is what it should have said.

Since there is an equal chance of finding the password in the first half of the total number of guesses as there is in finding it in the second half of the total number, it is accurate to assume that on average, a password will be found in 1/2 of the time required to do an exhaustive search.

> That means just over 11 months for the example just mentioned.

Author: ```suggestion
That means just over 5 months for the example just mentioned.
```
This is now correct given the revision to line 140
